### PR TITLE
[flink] Supports Kafka with canal CDC format ingestion

### PR DESCRIPTION
**Pull request by Alibaba-HZY**
### Purpose

Supports Kafka with canal CDC format.(#911)

### Tests

KafkaCanalSyncDatabaseActionITCase
KafkaCanalSyncTableActionITCase

### CODE REVIEW

File: docs/content/how-to/cdc-ingestion.md [hunk 1]
```diff
@@ -157,3 +157,187 @@ Example
...
+* `--catalog-conf` is the configuration for Paimon catalog. Each configuration should be specified in the format `key=value`. See [here]({{< ref "maintenance/configurations" >}}) for a complete list of catalog configurations.
+* `--table-conf` is the configuration for Paimon table sink. Each configuration should be specified in the format `key=value`. See [here]({{< ref "maintenance/configurations" >}}) for a complete list of table configurations.
+
+If the Paimon table you specify does not exist, this action will automatically create the table. Its schema will be derived from all specified Kafka topic's tables. If the Paimon table already exists, its schema will be compared against the schema of all specified Kafka topic's tables.
```
Comment (JingsongLi) [inline docs/content/how-to/cdc-ingestion.md:237]: >  Its schema will be derived from all specified Kafka topic's tables

How to infer schema from topic, we can document this.

File: docs/content/how-to/cdc-ingestion.md [hunk 2]
```diff
@@ -157,3 +157,187 @@ Example
...
+* `--catalog-conf` is the configuration for Paimon catalog. Each configuration should be specified in the format `key=value`. See [here]({{< ref "maintenance/configurations" >}}) for a complete list of catalog configurations.
+* `--table-conf` is the configuration for Paimon table sink. Each configuration should be specified in the format `key=value`. All Paimon sink table will be applied the same set of configurations. See [here]({{< ref "maintenance/configurations" >}}) for a complete list of table configurations.
+
+Only tables with primary keys will be synchronized.
+
+For each Kafka topic's table to be synchronized, if the corresponding Paimon table does not exist, this action will automatically create the table. Its schema will be derived from all specified Kafka topic's tables. If the Paimon table already exists, its schema will be compared against the schema of all specified Kafka topic's tables.
```
Comment (JingsongLi) [inline docs/content/how-to/cdc-ingestion.md:310]: > Its schema will be derived from all specified Kafka topic's tables.

How to infer schema from topic, we can document this.

File: docs/content/how-to/cdc-ingestion.md [hunk 3]
```diff
@@ -157,3 +157,187 @@ Example
...
+
```
Comment (JingsongLi) [inline docs/content/how-to/cdc-ingestion.md:311]: Do not support one topic multiple tables? What happen in this case? Document this.

**Review by JingsongLi — COMMENTED** [review]
Thanks @Alibaba-HZY  very much, very well done. Left some comments.
If possible, break out some PRs to speed up the review.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.apache.paimon.utils.Preconditions.checkArgument;
+
+/** {@link EventParser} for Canal-json. */
+public class CanalJsonEventParser implements EventParser<String> {
```
[the remark this answers is not in the document]
Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java:66]: ok i will submit another pr to test this class

[the remark this answers is not in the document]
Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/canal/CanalJsonEventParser.java:66]: > Can we have test for this class? Maybe you can create a separate PR for this to accelerate review.
KafkaEventParserTest in this pr 
https://github.com/apache/incubator-paimon/pull/1221

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSyncTableAction.java [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+ *             longer length,
+ *         <li>altering from an integer type (tinyint, smallint, int, bigint) to another integer
+ *             type with wider range,
+ *         <li>altering from a floating-point type (float, double) to another floating-point type
+ *             with wider range,
+ *       </ul>
+ *       are supported. Other type changes will cause exceptions.
+ * </ul>
+ */
+public class KafkaSyncTableAction implements Action {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSyncTableAction.java:87]: We can create a PR for `KafkaSyncTableAction` now~

Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSyncTableAction.java:87]: https://github.com/apache/incubator-paimon/pull/1239

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/TypeUtil.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.paimon.flink.action.cdc.kafka;
+
+import java.util.Arrays;
+import java.util.List;
+
+/** Get precision or scale from string type. */
+public class TypeUtil {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/TypeUtil.java:25]: no use?

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/cdc/kafka/KafkaActionITCaseBase.java [hunk 7]
```diff
@@ -79,11 +79,11 @@
 import static org.assertj.core.api.Assertions.assertThat;
 
 /** Base test class for {@link org.apache.paimon.flink.action.Action}s related to MySQL. */
-public class KafkaCanalActionITCaseBase extends ActionITCaseBase {
+public class KafkaActionITCaseBase extends ActionITCaseBase {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/cdc/kafka/KafkaActionITCaseBase.java:82]: abstract?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/TypeUtil.java (hunk 6 above)
Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/TypeUtil.java]: yes i have delete it

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSyncDatabaseAction.java [hunk 8]
```diff
@@ -192,13 +192,14 @@ public void build(StreamExecutionEnvironment env) throws Exception {
                         incompatibleMessage(table.schema(), mySqlSchema, identifier);
                 if (shouldMonitorTable(table.schema(), fromMySql, errMsg)) {
                     monitoredTables.add(mySqlSchema.tableName());
+                    fileStoreTables.add(table);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/mysql/MySqlSyncDatabaseAction.java:195]: Please rebase master.

**Comment by JingsongLi** [comment]
See https://github.com/apache/incubator-paimon/pull/1281 to modify canal parser too?

**Comment by Alibaba-HZY** [comment]
> See #1281 to modify canal parser too?

done

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSchema.java [hunk 9]
```diff
@@ -41,55 +42,32 @@
...
 /** Utility class to load canal kafka schema. */
 public class KafkaSchema {
 
     private static final int MAX_RETRY = 100;
+    private static final int MAX_READ = 1000;
 
-    private final ObjectMapper objectMapper = new ObjectMapper();
+    private static final ObjectMapper objectMapper = new ObjectMapper();
 
     private String databaseName;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSchema.java:57]: final

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSchema.java [hunk 10]
```diff
@@ -41,55 +42,32 @@
...
 import java.util.UUID;
 
 /** Utility class to load canal kafka schema. */
 public class KafkaSchema {
 
     private static final int MAX_RETRY = 100;
+    private static final int MAX_READ = 1000;
 
-    private final ObjectMapper objectMapper = new ObjectMapper();
+    private static final ObjectMapper objectMapper = new ObjectMapper();
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/cdc/kafka/KafkaSchema.java:55]: why static again?

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**